### PR TITLE
[FSTORE-328] throw intelligible error message if user set primary and partition keys doesn't exist

### PR DESCRIPTION
**Pull request by davitbzh**
This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

**Review by moritzmeister — CHANGES_REQUESTED** [review]
Please write unit tests.

**Review by SirOibaf — CHANGES_REQUESTED** [review]
Add Java tests as well please

File: java/src/test/java/com/logicalclocks/hsfs/TestFeatureGroup.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/*
+ *  Copyright (c) 2022. Logical Clocks AB
```
Comment (SirOibaf) [inline java/src/test/java/com/logicalclocks/hsfs/TestFeatureGroup.java:2]: Hopsworks AB

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupUtils.java [hunk 2]
```diff
@@ -305,4 +305,30 @@ public Schema getDeserializedAvroSchema(String avroSchema) throws FeatureStoreEx
       throw new FeatureStoreException("Failed to deserialize online feature group schema" + avroSchema + ".");
     }
   }
+
+  public void verifyAttributeKeyNames(String primaryOrPartition, List<String> primaryPartitionKeyNames,
+                                      String eventTime, String precombineKeyName,
+                                      List<Feature> features) throws FeatureStoreException {
```
Comment (SirOibaf) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupUtils.java:311]: I think this method has 2 problems:
- The code is condensed in a single method, but we need to call the method 4 times to validate all the input parameters. I think we can do the validation in a single pass.
- The method is using strings and things being null for the method to have different behaviors. This is usually a recipe for disaster. Someone somewhere will call this method with the wrong parameters and either won't fully validate the feature group or the method will blow up.

I think we should either break down the method in 4 different methods, or do the validation in a single step. I prefer the second.

File: java/src/main/java/com/logicalclocks/hsfs/engine/ExternalFeatureGroupEngine.java [hunk 3]
```diff
@@ -40,6 +40,10 @@ public ExternalFeatureGroup saveFeatureGroup(ExternalFeatureGroup externalFeatur
           externalFeatureGroup.getTimeTravelFormat()));
     }
 
+    // verify primary keys
+    utils.verifyAttributeKeyNames("primary", externalFeatureGroup.getPrimaryKeys(),
```
Comment (SirOibaf) [inline java/src/main/java/com/logicalclocks/hsfs/engine/ExternalFeatureGroupEngine.java:44]: External feature groups can also have event time.

File: python/hsfs/util.py [hunk 4]
```diff
@@ -220,6 +221,28 @@ def get_hostname_replaced_url(sub_path: str):
     return url_parsed.geturl()
 
 
+def verify_attribute_key_names(attribute_name, provided_names, features):
```
Comment (SirOibaf) [inline python/hsfs/util.py:224]: Same comment as for the java method. The method has all the code but needs to be called several times to fully validate the feature group.

File: python/hsfs/core/external_feature_group_engine.py [hunk 5]
```diff
@@ -31,6 +31,9 @@ def save(self, feature_group):
 
         # set primary and partition key columns
         # we should move this to the backend
+        util.verify_attribute_key_names(
+            "primary", feature_group.primary_key, feature_group.features
+        )
```
Comment (SirOibaf) [inline python/hsfs/core/external_feature_group_engine.py:36]: Same comment as for the java part - external feature groups can also have event time columns

**Review by SirOibaf — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupUtils.java [hunk 6]
```diff
@@ -305,4 +305,41 @@ public Schema getDeserializedAvroSchema(String avroSchema) throws FeatureStoreEx
...
+          + " doesn't exist in feature dataframe");
+    }
+
+    if (featureGroup.getEventTime() != null && !featureNames.contains(featureGroup.getEventTime())) {
+      throw new FeatureStoreException("Provided eventTime feature name " + featureGroup.getEventTime()
+          + " doesn't exist in feature dataframe");
+    }
+  }
+
+
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupUtils.java:333]:
```suggestion
```

File: python/hsfs/util.py [hunk 7]
```diff
@@ -220,6 +221,34 @@ def get_hostname_replaced_url(sub_path: str):
     return url_parsed.geturl()
 
 
+def verify_attribute_key_names(feature_group_obj):
+    feature_names = set([feat.name for feat in feature_group_obj.features])
```
Comment (moritzmeister) [inline python/hsfs/util.py:225]:
```suggestion
    feature_names = set(feat.name for feat in feature_group_obj.features)
```
unnecessary to convert between the two

File: python/hsfs/util.py [hunk 8]
```diff
@@ -220,6 +221,34 @@ def get_hostname_replaced_url(sub_path: str):
...
+    if feature_group_obj.primary_key:
+        diff = list(set(feature_group_obj.primary_key) - feature_names)
```
Comment (moritzmeister) [inline python/hsfs/util.py:227]:
```suggestion
        diff = set(feature_group_obj.primary_key) - feature_names
```

File: python/hsfs/util.py [hunk 9]
```diff
@@ -220,6 +221,34 @@ def get_hostname_replaced_url(sub_path: str):
...
+        if diff:
+            raise exceptions.FeatureStoreException(
+                f"Provided primary key(s) {','.join(diff)} doesn't exist in feature dataframe"
+            )
+    if feature_group_obj.partition_key:
+        diff = list(set(feature_group_obj.partition_key) - feature_names)
```
Comment (moritzmeister) [inline python/hsfs/util.py:233]:
```suggestion
        diff = set(feature_group_obj.partition_key) - feature_names
```

**Review by moritzmeister — CHANGES_REQUESTED** [review]

**Review by moritzmeister — APPROVED** [review]

**Comment by moritzmeister** [comment]
@davitbzh please rebase on master so that unit tests pass.

**Review by SirOibaf — APPROVED** [review]

**Pull request merged**